components/builder-api/src/server/resources/origins.rs

```diff
@@ -269,15 +272,108 @@ fn delete_origin(req: HttpRequest, path: Path<String>, state: Data<AppState>) ->
+                Ok(_) => HttpResponse::NoContent().into(),
+                Err(err) => {
+                    debug!("Origin {} deletion failed! err = {}", origin, err);
+                    // We do not want to expose any database details from diesel
+                    // thus we simply return a 409 with an empty body.
+                    HttpResponse::new(StatusCode::CONFLICT)
+                }
+            }
+        }
         Err(err) => {
-            debug!("Origin {} is not deletable, err = {}", origin, err);
-            HttpResponse::new(StatusCode::CONFLICT)
+            debug!("Origin preflight determined that {} is not deletable, err = {}!",
+                   origin, err);
+            // Here we want to enrich the http response with a sanitized error
+            // by returning a 409 with a helpful message in the body.
+            HttpResponse::with_body(StatusCode::CONFLICT, Body::from_message(format!("{}", err)))
         }
     }
 }
 
+#[allow(clippy::needless_pass_by_value)]
+fn origin_delete_preflight(origin: &str, conn: &PgConnection) -> Result<()> {
+    match Project::count_origin_projects(&origin, &*conn) {
+        Ok(0) => {}
+        Ok(count) => {
+            let err = format!("There are {} projects remaining in origin {}. Must be zero.",
+                              count, origin);
+            return Err(Error::BuilderCore(OriginDeleteError(err)));
+        }
+        Err(e) => return Err(Error::DieselError(e)),
+    };
+
+    match OriginMember::count_origin_members(&origin, &*conn) {
+        // allow 1 - the origin owner
+        Ok(1) => {}
+        Ok(count) => {
+            let err = format!("There are {} members remaining in origin {}. Only one is allowed.",
+                              count, origin);
+            return Err(Error::BuilderCore(OriginDeleteError(err)));
+        }
+        Err(e) => {
+            return Err(Error::DieselError(e));
+        }
+    };
+
+    match OriginSecret::count_origin_secrets(&origin, &*conn) {
+        Ok(0) => {}
+        Ok(count) => {
+            let err = format!("There are {} secrets remaining in origin {}. Must be zero.",
+                              count, origin);
+            return Err(Error::BuilderCore(OriginDeleteError(err)));
+        }
+        Err(e) => {
+            return Err(Error::DieselError(e));
+        }
+    };
+
+    match OriginIntegration::count_origin_integrations(&origin, &*conn) {
+        Ok(0) => {}
+        Ok(count) => {
+            let err = format!("There are {} integrations remaining in origin {}. Must be zero.",
+                              count, origin);
+            return Err(Error::BuilderCore(OriginDeleteError(err)));
+        }
+        Err(e) => {
+            return Err(Error::DieselError(e));
+        }
+    };
+
+    match Channel::count_origin_channels(&origin, &*conn) {
+        // allow 2 - [unstable, stable] channels cannot be deleted
+        Ok(2) => {}
```

Comment:
Similar to the count above. If at some point we change how the channels work this could no longer be correct. I'll defer to others on if this is a concern for now though.

Reply:
I think this should be Ok for now, but yeah with the channel refactor it might need to change. I don't imagine any change we make will massively complicate this.